torch/testing/_internal/common_utils.py

```diff
@@ -940,6 +940,35 @@ def get_comparison_dtype(a, b):
 
     return compare_dtype
 
+def compare_with_reference(torch_fn, ref_fn, sample_input):
+    # Compares torch function with reference function for given sample input (object of SampleInput)
+    # Note: only values are compares, type comparison is not done here
+    try:
+        n_inp, n_args, _ = sample_input.numpy()
+    except TypeError:
+        # TODO: This needs a comment
+        return
+    t_inp, t_args, _ = sample_input.splat()
+
+    t_kwargs, n_kwargs = torch_fn.sample_kwargs(t_inp.device, None, t_inp)
+    actual = torch_fn(t_inp, *t_args, **t_kwargs)
+    expected = ref_fn(n_inp, *n_args, **n_kwargs)
+
+    # Crafts a custom error message for smaller, printable tensors
+    if t_inp.numel() < 10:
+        msg = ("Failed to produce expected results! Input tensor was"
+               " {0}, torch result is {1}, and reference result is"
+               " {2}.").format(t_inp, actual, expected)
+    else:
+        msg = None
+
+    if isinstance(actual, torch.Tensor):
+        torch.testing.assert_close(actual, torch.as_tensor(expected, dtype=actual.dtype), msg=msg)
+    else:
+        for x, y in zip(expected, actual):
+            # Testing multi-outputs results
+            torch.testing.assert_close(actual, expected, msg=msg)
```

Comment:
Note to self: This needs comment.

Note to reviewers: As per discussion, and also offline discussion with @pmeier - we want to limit this reference check (for now at least) to only test values and not types.

Reply:
Update: a `check_device=False` param has been added here to ensure no device checks are done for reference tests.